Added preinstall hooks (pip install and shell scripts)

jaklinger opened the pull request:
Closes #20 

**Just a note up front: 11 out of 17 of the changed files are relating to testing so don't panic 😄 **

Introduces the following features:

* Any packages listed in the config under `PREINSTALL_PKGS` are installed for all batch tasks, so long as they are in the local environment. Packages are skipped if they're not in the local environment, meaning that switching off his behaviour is as simple as overwriting these variables.
* If script(s) `preinstall-{flow_name}.sh` and/or `preinstall-{flow_name}-{step_name}.sh` are present then they are executed prior to batch step execution.
* `daps-utils` and `mysql-connector-python` are defaults for `PREINSTALL_PKGS` as we would need them for all daps implementations at present. Happy to move this to daps only repos if preferred, but I would reiterate that these won't be installed unless `daps-utils` is already installed in the user's local env, which would kind of imply that you wanted to bundle these.

Originally I had intended that in the presence of `environment=conda` that `PREINSTALL_PKGS` wouldn't be installed however this wouldn't actually work in practice as e.g. `daps-utils` would still be needed at execution time. It is possible to implement this feature if you really think that it's desirable (it's just a few lines extra). Otherwise one minor workaround would be to toggle `PREINSTALL_PKGS` and use separate flows for packages that require `PREINSTALL_PKGS` and conda.

Additional integration tests
============================

Integration tests have been added for `batch_flow.py` and `batch_flow_with_conda.py` (described below). One caveat with this is that it increases test run time to 20 minutes(!) because batch tasks are submitted. One way to resolve this would be to set up a separate test matrix to allow for parallel testing, if that would be desirable?

How to check
============

* Make sure `daps-utils` and `mysql-connector-python` are installed in your working environment
* From `metaflow_extensions/tests/myproject/myproject/flows` run the following:

Integration 1: batch
--------------------

The following flow demonstrates that `daps-utils` is preinstalled also that the `preinstall-batch_flow.sh` script has run in the batch steps, both task prior to execution.

```
PYTHONPATH=~/Nesta/metaflow_extensions:~/Nesta/metaflow_extensions/tests/myproject python batch_flow.py --datastore=s3 run
```


Integration 2: batch + conda
----------------------------

The following flow demonstrates that `daps-utils` is preinstalled also that the `preinstall-batch_flow.sh` script has run in the batch steps, both task prior to execution. The special case here is that metaflow uses an additional set of pre-install mechanisms with conda and so the implementation is a little more involved.

```
PYTHONPATH=/path/to/metaflow_extensions:/path/to/metaflow_extensions/tests/myproject python batch_flow_with_conda.py --datastore=s3 --environment=conda run
```

bishax commented on metaflow_extensions/plugins/install_step_decorator.py:
```diff
@@ -13,17 +13,88 @@
+        # project_root = up_to_project_root(flow_path)
+        # if project_root:
+        #     pip_install(python_exec, project_root)
+        # return env_id
```
Just went to test whether I could get things to break where `myproject` imports unavailable dependencies and in doing so realised that the return statement was accidentally commented out which borks things.

bishax replied:
This is also why the CI tests are failing.

And commenting the above logic as per my suggestion breaks everything because that logic (duplicated from `task_pre_step`) has to be run earlier because of `daps_utils` weird backwards dependency on `myproject`

jaklinger replied:
(sorry should have said that I'm still investigating, I've got half a report together on the full execution sequence to disentangle things)

jaklinger replied:
I would say that although the backwards dependency of `daps_utils` on `myproject` is weird, it's not particularly weird to expect the current project to be available at runtime, i.e. what `daps_utils` expects is that the following would be possible - which it currently isn't in with `environment=conda`:

```python
import myproject
class MyFlow(FlowSpec):
    ...
```

the duplication of the snippet from `task_pre_step` basically makes this possible, which I think is pretty reasonable isn't it?

bishax replied:
> which I think is pretty reasonable isn't it?
I guess so, because `myproject` isn't just another dependency like you would pass to `@conda`.

jaklinger commented on metaflow_extensions/plugins/install_step_decorator.py:
```diff
@@ -15,15 +15,55 @@
+        flow_path = Path(sys.argv[0])
+        project_root = up_to_project_root(flow_path)
+        if project_root:
+            pip_install(python_exec, project_root)
```
A took me a little while to get to the bottom of this. The flow lifecycle for non-conda environments looks like:

```
1) ProjectEnvironment.decospecs

2) ProjectEnvironment.add_to_package

3) for each step:
      b) Run(flow)      # <-- {myproject} not installed yet, but in `PYTHONPATH`
      b) Start(step) 
      c) ProjectEnvironment.task_pre_step
```

whereas for `environment=conda` looks like:

```
1) CondaEnvironment.decospecs --> ProjectEnvironment.decospecs

2) for each step:
      CondaStepDecorator._prepare_step_environment

3) ProjectEnvironment.add_to_package

4) for each step:
      a) CondaStepDecorator._prepare_step_environment   (again!)
      b) Run(flow)      # <-- {myproject} not installed yet
      b) Start(step) 
      c) ProjectEnvironment.task_pre_step
```

and so installing the `{myproject}` in `_prepare_step_environment` is one way in which to bring it into scope prior to `Run(flow)`. I've at least refactored this snippet into `utils` to at least reduce the minor boilerplate.

jaklinger commented on tests/utils.py:
```diff
@@ -62,8 +62,15 @@ def run_flow(
+    # daps projects need to be able to import {myproject}
+    # which we can figure out from the executable in this test
+    env = dict(os.environ)
+    env.update({"PYTHONPATH": path.parent.parent.parent})
```
https://github.com/nestauk/metaflow_extensions/pull/21#discussion_r740181752

jaklinger commented on metaflow_extensions/utils.py:
```diff
@@ -53,38 +60,56 @@ def running_in_local_env() -> bool:
-def upgrade_pip() -> None:
+def local_pip_install(project_path: Path) -> None:
+    """`pip install` local package at `project_path`."""
+    pip_install(sys.executable, project_path)
```
Added min version into reqs

jaklinger commented:
Ok, so thanks again for the comments @bishax - things a looking a lot slicker now. A minor changelog since the last review:

1. Removed `sh` from everything under `metaflow_extensions/**` and falls back on `subprocess` (and no more grep (!))
2. Changed the weird sha in `.precommit_.[].yaml` to `v2.4.1` of the mirror repo
3. Better docstrings to help guide future developers
4. The pip install duplication that we've discussed has been refactored into utils
5. A minimum version has been set on `pip` to allow dropping of the `in-tree-build` arg/warning.

Just waiting on the tests then I'll mark for re-review

bishax commented on metaflow_extensions/plugins/install_step_decorator.py:
```diff
@@ -11,33 +11,61 @@
+    bootstrap_commands method is executed immediately prior to the batch runtime,
+    and is never called locally.
+
+    In short, the difference between install_step_decorator.prepare_step_wrapper
```
I think this means `install_step_decorator.task_pre_step`?

bishax commented on metaflow_extensions/plugins/install_step_decorator.py:
```diff
@@ -11,33 +11,61 @@
+    In the flow lifecycle this is executed prior to steps being explicitly executed:
+    in practice this occurs in conda_step_decorator.package_init, and it is only
+    executed on local steps (i.e. not batch steps). There is some duplication with
+    the prepare_step_wrapper method in the install_step_decorator module as the
```
`s/prepare_step_wrapper/task_pre_step`?

bishax commented on metaflow_extensions/plugins/install_step_decorator.py:
```diff
@@ -49,11 +77,8 @@ def wrapped_prepare_step_environment(self, *args, **kwargs):
+        install_flow_project(python_exec)
         return env_id
 
     return wrapped_prepare_step_environment
```
Lines 87-89 (can't select those)...

Should the patching be protected by some kind of check?

We probably don't want to patch if `METAFLOW_DEFAULT_ENVIRONMENT` isn't `project`?

The behaviour for installing `PREINSTALL_PKGS` is coupled to the "project" environment, `@install_project` can actually act independently of the "project" environment.

This also suggests that this funcitonality should be moved to `project_environment.py`?

jaklinger replied:
yep, fair, there was little reason in the end to house this in `install_step_decorator` at all, just an early design decision that stuck

bishax commented on metaflow_extensions/utils.py:
```diff
@@ -60,50 +54,94 @@ def running_in_local_env() -> bool:
+    subprocess_output: subprocess.CompletedProcess,
+) -> Iterator[str]:
+    """Split and decode subprocess output."""
+    yield from map(bytes.decode, subprocess_output.stdout.splitlines())
```
:heart:

bishax reviewed (commented):
A few su-questions on the docstring (apologies for being picky on this but want to make sure it's easy to recall how/why things work when this code is looked at down the line.

The only other thing was a suggestion to move functionality from `install_step_decorator.py` to `project_environment.py` as `@install_project` can actually function independently from the "project" environment (whose behaviour is changing in this PR).

After that GTM

jaklinger commented on metaflow_extensions/plugins/install_step_decorator.py:
```diff
@@ -11,33 +11,61 @@
+    In short, the difference between install_step_decorator.prepare_step_wrapper
+    and project_environment.bootstrap_wrapper is that:
+
+      * prepare_step_wrapper patches the local conda environment only (it is unable to
+        patch the batch conda environment because different of architectures, e.g.
+        MacOS on local vs Linux on batch)
```
indeed, although the difference in architectures will explicitly cause a crash on lines 78 (`pip_install(python_exec, pkg_spec)`) and 81 (`install_flow_project(python_exec)`) 🙂 if submitting from OSX as opposed to linux. But as you say that patching is inconsequential for batch anyway. I'll add that in

bishax commented on metaflow_extensions/plugins/project_environment.py:
```diff
@@ -73,3 +178,28 @@ def add_to_package(self):
+        from metaflow_extensions.config.metaflow_config import PREINSTALL_PKGS
+
+        # Run any pre-install scripts
+        flow_name = Path(sys.argv[0]).stem
+        preinstalls = (
+            "preinstall.sh",
+            f"preinstall-{flow_name}.sh",
+            f"preinstall-{flow_name}-{step_name}.sh",
+        )
+        preinstalls = list(filter(lambda fname: Path(fname).exists(), preinstalls))
```
Just come across an error, need to incorporate the flow directory into the path resolution - the current working directory may not be the same as the flow directory....

```suggestion
        flow_name = Path(sys.argv[0]).stem
        flow_dir = Path(sys.argv[0]).parent
        preinstalls = (
            "preinstall.sh",
            f"preinstall-{flow_name}.sh",
            f"preinstall-{flow_name}-{step_name}.sh",
        )
        preinstalls = list(
            filter(lambda fname: (flow_dir / fname).exists(), preinstalls)
        )
```

jaklinger commented:
Minor changelog since last review:

* step decorator moved to `project_environment`, with logic to avoid patching if the environment isn't "project"
* docstring clarification
* included your suggestion just now on including the flow directory for the pre-installs

bishax commented:
@jaklinger I've just pushed two commits:
1) Fixing the bug discussed on Slack
2) I ran mypy and discovered a few type errors

Feel free to revert / edit.

If you don't think those last 2 commits need changes them good to merge.
